export_imaqal_ad_contacts.py

```diff
@@ -0,0 +1,120 @@
+    parser.add_argument("firebase_credentials_file_url", metavar="firebase-credentials-file-url",
+                        help="GS url to the credentials file to use to access the uuid table")
+    parser.add_argument("uuid_table_name", metavar="uuid-table-name",
+                        help="Name of the uuid table to use to re-identify the avf ids")
+    parser.add_argument("analysis_csv_paths", metavar="analysis-csv-paths", nargs="+",
+                        help="Paths to the analysis csv files (either messages or individuals) to extract phone "
+                             "numbers from")
+    parser.add_argument("csv_output_file_path", metavar="csv-output-file-path",
+                        help="Path to a CSV file to write the contacts from the locations of interest to. "
+                             "Exported file is in a format suitable for direct upload to Rapid Pro")
+
+    args = parser.parse_args()
+
+    exclusion_list_file_path = args.exclusion_list_file_path
+    google_cloud_credentials_file_path = args.google_cloud_credentials_file_path
+    firebase_credentials_file_url = args.firebase_credentials_file_url
+    uuid_table_name = args.uuid_table_name
+    analysis_csv_paths = args.analysis_csv_paths
+    csv_output_file_path = args.csv_output_file_path
+
+    log.info("Downloading Firestore UUID Table credentials...")
+    firestore_uuid_table_credentials = json.loads(google_cloud_utils.download_blob_to_string(
+        google_cloud_credentials_file_path,
+        firebase_credentials_file_url
+    ))
+
+    phone_number_uuid_table = FirestoreUuidTable(
+        uuid_table_name,
+        firestore_uuid_table_credentials,
+        "avf-phone-uuid-"
+    )
+    log.info("Initialised the Firestore UUID table")
+
+    all_uuids = set()
+    banadir_uuids = set()
+    sws_uuids = set()
+    for path in analysis_csv_paths:
+        # Load the traced data
+        log.info(f"Loading analysis file '{path}'...")
+        with open(path) as f:
+            data = list(csv.DictReader(f))
+        log.info(f"Loaded {len(data)} rows")
+
+        for row in data:
+            if row["consent_withdrawn"] == Codes.TRUE:
+                continue
+
+            if row["state"] == SomaliaCodes.BANADIR:
+                banadir_uuids.add(row["uid"])
+
+            if row["state"] == SomaliaCodes.SOUTH_WEST_STATE:
+                sws_uuids.add(row["uid"])
+
+            all_uuids.add(row["uid"])
+    log.info(f"Loaded {len(all_uuids)} uuids ({len(banadir_uuids)} banadir, {len(sws_uuids)} sws)")
+
+    # Take a random sample of 15% of participants from each location
+    sampled_banadir = random.sample(banadir_uuids, int(0.15 * len(banadir_uuids)))
+    sampled_sws = random.sample(sws_uuids, int(0.15 * len(sws_uuids)))
+    log.info(f"Sampled {len(sampled_banadir)}/{len(banadir_uuids)} banadir contacts")
+    log.info(f"Sampled {len(sampled_sws)}/{len(sws_uuids)} sws contacts")
+
+    uuids = sampled_banadir + sampled_sws
```

Comment:
filtering out the exclusion list after sampling will reduce the sample size further. I would suggest we exclude before sampling.

Reply:
moved the exclusion list filtering to take place before sampling